pysynchronizer/storage.py

```
@@ -0,0 +1,89 @@
+#
+
+
+import uuid
+
+from os import O_CREAT, O_WRONLY
+from os.path import join, split
+from .http_fetcher import HttpFetcher
+from .icbinn import Icbinn
+
+DOWNLOAD_BLOCK_SIZE = 512 * 1024
+REPO_DOWNLOAD_DIR = 'repo-download'
+REPO_HANDOVER_DIR = 'repo'
+DISKS_DIR = 'disks'
+
+class Storage:
+    def __init__(self, block_size=DOWNLOAD_BLOCK_SIZE):
+        self.storage = Icbinn().storage
+        self.block_size = block_size
+
+    def fetch_using_partial(self, url, file_path):
+        partial = file_path + '.partial'
+        offset = 0
+        fetcher = HttpFetcher()
+
+        if self.storage.exists(partial):
+            offset = self.storage.stat(partial)[0]
+
+        partial_handle = self.storage.open(partial, O_CREAT | O_WRONLY)
+        if offset != 0:
+            partial_handle.seek(offset)
+
+        fetcher.stream(url, partial_handle, offset, self.block_size)
+
+        partial_handle.close()
+
+        if self.storage.exists(file_path):
+            self.storage.unlink(file_path)
+
+        self.storage.rename(partial, file_path)
+
+    def download_disk(self, disk_name, url):
+        disk_file = join(DISKS_DIR, disk_name)
+
+        self.fetch_using_partial(url, disk_file)
+
+        return disk_file
+
+    def list_disks(self):
+        return self.storage.listdir(DISKS_DIR)
+
+    def stage_oxt_repo(self, url):
+        repo_uuid = uuid.uuid5(uuid.NAMESPACE_URL, url)
+        repo_name = str(repo_uuid) + '.tar'
+
+        download_file = join(REPO_DOWNLOAD_DIR, repo_name)
+
+        self.fetch_using_partial(url, download_file)
+
+        return repo_name
+
+    def list_repo_stage(self):
+        return self.storage.listdir(REPO_DOWNLOAD_DIR)
+
+    def apply_oxt_repo(self, name):
+        src = join(REPO_DOWNLOAD_DIR, name)
+        dst = join(REPO_HANDOVER_DIR, name)
+
+        if self.storage.exists(dst):
+            return False
+
+        self.storage.makedirs(REPO_HANDOVER_DIR)
```

I thought I made this comment already, but it seems to have been lost?

This triggers an EEXIST error in icbinn_svc, so maybe something like the untested:
```
if ! self.storage.exists(REPO_HANDOVER_DIR):
    self.storage.makedirs(REPO_HANDOVER_DIR)
```

This would silence the error at least.  However, REPO_HANDOVER_DIR needs to be labeled updatemgr_sync_client_storage_t, so icbinn_svc today can't create it.  A filetrans would help there.

This code may also rely on REPO_DOWNLOAD_DIR already existing.

---

I am going back and forth on this one. Does the check belong in IcbinnRemote.makedirs or should every caller be responsible for doing the existence check.

---

Looking more at IcbinnRemote.makedirs, IMHO it makes more sense to put the check there.

---

Moving the check into IcbinnRemote.makedirs is good, but stage_oxt_repo should gain `self.storage.mkdir(REPO_DOWNLOAD_DIR)`.

---

I think you mean self.storage.makedirs(), and yes, will add.